// Data.String/index.js
"use strict";
function charAt(i) {  return function(s) {    return s.charAt(i);   };};
function charCodeAt(i) {  return function(s) {    return s.charCodeAt(i);   };};
function fromCharCode(n) {  return String.fromCharCode(n);};
function indexOf(x) {  return function(s) {    return s.indexOf(x);  }; };
function indexOf$prime(x) {  return function(startAt) {    return function(s) {      return s.indexOf(x, startAt);    };   }; };
function lastIndexOf(x) {  return function(s) {    return s.lastIndexOf(x);  };};
function lastIndexOf$prime(x) {  return function(startAt) {    return function(s) {      return s.lastIndexOf(x, startAt);    };   }; };
function length(s) {  return s.length;};
function localeCompare(s1) {  return function(s2) {    return s1.localeCompare(s2);  };};
function replace(s1) {  return function(s2) {    return function(s3) {      return s3.replace(s1, s2);    };  };};
function take(n) {  return function(s) {    return s.substr(0, n);  };};
function drop(n) {  return function(s) {    return s.substr(n);  };};
function split(sep) {  return function(s) {    return s.split(sep);  };};
function toLower(s) {  return s.toLowerCase();};
function toUpper(s) {  return s.toUpperCase();};
function trim(s) {  return s.trim();};
function joinWith (s) {  return function (xs) {    return xs.join(s);  };};
module.exports = {
    joinWith: joinWith, 
    trim: trim, 
    toUpper: toUpper, 
    toLower: toLower, 
    split: split, 
    drop: drop, 
    take: take, 
    replace: replace, 
    localeCompare: localeCompare, 
    length: length, 
    "lastIndexOf'": lastIndexOf$prime, 
    lastIndexOf: lastIndexOf, 
    "indexOf'": indexOf$prime, 
    indexOf: indexOf, 
    fromCharCode: fromCharCode, 
    charCodeAt: charCodeAt, 
    charAt: charAt
};
